File: java-apps/kq-aggregate/src/main/java/com/kafkaquest/kq/aggregate/services/handlers/EventHandlerFactory.java
package com.kafkaquest.kq.aggregate.services.handlers;

import com.kafkaquest.kq.common.util.models.kafkakeys.ApiKey;

public class EventHandlerFactory {

    public static EventHandler getHandler(ApiKey apiKey) {

        return switch (apiKey.getEventType()) {
            case GAME_STARTED_EVENT -> new GameStartedEventHandler();
            case PIECE_ADDED_EVENT -> new PieceAddedEventHandler();
            case PIECE_REMOVED_EVENT -> new PieceRemovedEventHandler();
            case REVERT_MOVES_EVENT -> new RevertMovesEventHandler();
            case DISCARD_GAME_EVENT -> new DiscardGameEventHandler();
            default -> throw new IllegalArgumentException("Unknown Event");
        };
    }
}

File: kq-projection-dotnet/kq-projection/Services/EventHandling/EventHandlerFactory.cs
using KQ_Projection.Common.Enums;
using KQ_Projection.Common.Models.Keys;

namespace KQ_Projection.Services.EventHandling;

public class EventHandlerFactory
{
    public static EventHandler GetHandler(AggregateKey key) {
        return key.eventType switch {
            EventType.GAME_STARTED_EVENT => new GameStartedEventHandler(),
            EventType.PIECE_ADDED_EVENT => new PieceAddedEventHandler(),
            EventType.PIECE_REMOVED_EVENT => new PieceRemovedEventHandler(),
            EventType.EXCEPTIONAL_EVENT => new ExceptionalEventHandler(),
            EventType.DISCARD_GAME_EVENT => new DiscardGameEventHandler(),
            _ => throw new ArgumentNullException(nameof(key))
        };
    }
}
